libimage/runtime.go

```diff
@@ -574,23 +574,30 @@ func (r *Runtime) ListImages(ctx context.Context, names []string, options *ListI
 		options = &ListImagesOptions{}
 	}
 
-	var images []*Image
-	if len(names) > 0 {
-		for _, name := range names {
-			image, _, err := r.LookupImage(name, nil)
-			if err != nil {
-				return nil, err
-			}
-			images = append(images, image)
-		}
-	} else {
-		storageImages, err := r.store.Images()
+	lookupImagesMap := make(map[string]struct{}, 0)
+	for _, name := range names {
+		image, _, err := r.LookupImage(name, nil)
 		if err != nil {
 			return nil, err
 		}
-		for i := range storageImages {
-			images = append(images, r.storageToImage(&storageImages[i], nil))
+		lookupImagesMap[image.ID()] = struct{}{}
```

Comment:
🤔 

Note that this is racy: we can match `name` against id1, but by the time `MultiList` is called, `name` may now be matching `id2`, and we will not see any match at all.

At the very least, it would help future maintainers if this race were documented in a comment.

But, worse, many name changes (e.g. pulling `foo:latest` are _atomic_ in c/storage, where we atomically move the tag from one image to another; and this could cause even such atomic changes to result in “`name` not found”.

But, then again, we probably don’t want to turn `LookupImage` inside out to match on image names we get from `storageListResult`.

I don’t immediately know what to do here. A local fix seems hard, maybe changing the callers instead could be easier.

Reply:
My idea is this: 
1) Look for images by name.
2) Take a snapshot of the layers and images.
3) Filter the snapshot`s images by the ID of the image that was looked up in the previous step. 
     - If the ID is not found, it is not a problem because the image was removed between calls. 
     - If the image has been changed and does not match the name search. That's a problem of outdated data, so the caller could deal with that. (But it will have to check if there is any handling or if it doesn't break something.)
     
It minimizes the races.

Reply:
I assume the ID will not change when the name changes.

Reply:
Image _contents_ are immutable and the ID implies the same contents (notably the same layers); but image _names_ are mutable, and the same ID can have different names over time.
Consider
- pull A:latest: pulls ID1 = v1, names it A:latest
- pull A:latest: pulls ID2 = v2, names it A:latest; ID1 continues to exist, but the A:latest name is removed
- ListImages now looks for A:latest, finds ID2
- pull A:latest: pulls ID1 = v1 (because the image author has reverted the tag), which already exists; adds an A:latest name to ID1, removes it from ID2
- user completely removes the ID2 image
- ListImages now checks for the image ID2, and reports that A:latest was not found. That’s incorrect.

You’re right that if an image listing is happening concurrently with image pulls, the caller may get an image which no longer has the required name by the time this function returns — and that’s fine.

But the name changes by pull are _atomic_ in c/storage, after the initial pull A:latest _always_ points to an image. The user will never see an “A:latest not found” error. And that matters in practice: one way the pulls of the same name can happen is by `podman auto-update`.

Reply:
Okay, I have an idea how to solve this. MultiListOptions can be extended with a [callback function](https://github.com/Honny1/storage/blob/callback/store.go#L92). This callback function is executed after getting all data with Images and Layer locks. So name lookups could be done with Images and Layers locked, so all data would be consistent. 

@mtrmac WDYT?

Reply:
- (Conceptually this discussion shows what he c/storage PR should not have been merged until this PR demonstrates that we get everything we need … oh well, already happened now.)
- In general, c/storage exposing a callback that runs with locks held seems problematic:
  - That callback could trigger another c/storage operation and cause a deadlock
  - This would make the c/storage lock existence / structure an API; exposing an implementation detail like that just seems undesirable.
- Read some of the name lookup code … it does various c/storage lookups, and those _can’t_ possibly be called with c/storage locks already held. (And even restructuring all of that to run on a `MultiList` instead of dealing with c/storage directly seems fairly hard — doesn’t it use c/image/v5/storage to trigger lookups in some cases? Would that have to be restructured / parametrized to use a `MultiList`?